src/mc-range-encoding.c

```diff
@@ -313,3 +313,297 @@ mc_getTypeInfoDouble (mc_getTypeInfoDouble_args_t args,
+   }
+
+   // The coefficient of the number, without exponent/sign
+   const mlib_int128 coeff = mc_dec128_coeff (args.value);
+
+   if (mlib_int128_eq (coeff, i128_zero)) {
+      // If the coefficient is zero, the result is encoded as the midpoint
+      // between zero and 2^128-1
+      *out = (mc_OSTType_Decimal128){
+         .value = i128_2pow127,
+         .min = i128_zero,
+         .max = MLIB_INT128_UMAX,
+      };
+      return true;
+   }
+
+   // Coefficient is an unsigned value. We'll later scale our answer based on
+   // the sign of the actual Decimal128
+   const bool isNegative = mc_dec128_is_negative (args.value);
+
+   // cMax = 10^34 - 1 (The largest integer representable in Decimal128)
+   const mlib_int128 cMax =
+      mlib_int128_sub (mlib_int128_pow10 (34), MLIB_INT128_CAST (1));
+   const mlib_int128 cMax_div_ten = mlib_int128_div (cMax, i128_ten);
+
+   // The biased exponent from the decimal number. The paper refers to the
+   // expression (e - e_min), which is the value of the biased exponent.
+   const uint32_t exp_biased = mc_dec128_get_biased_exp (args.value);
+
+   // ρ (rho) is the greatest integer such that: coeff×10^ρ <= cMax
+   unsigned rho = 0;
+   // Keep track of the subexpression coeff×10^ρ rather than recalculating it
+   // time.
+   // Initially: (ρ = 0) -> (10^ρ = 1) -> (coeff×10^ρ = coeff×1 = coeff):
+   mlib_int128 coeff_scaled = coeff;
+   // Calculate ρ: This could be done using a log10 with a division, but that
+   // is far more work than just a few multiplications.
+   // While: coeff×ten^ρ < cMax/10:
+   while (mlib_int128_ucmp (coeff_scaled, cMax_div_ten) < 0) {
+      // Increase rho until we pass cMax/10
+      rho++;
+      // Scale our computed subexpression rather than fully recomputing it
+      coeff_scaled = mlib_int128_mul (coeff_scaled, i128_ten);
+   }
+
+   // No multiplication by 10 should ever send us from N < cMax/10 to N > cMax
+   BSON_ASSERT (mlib_int128_ucmp (coeff_scaled, cMax) <= 0);
+
+   mlib_int128 result;
+   if (rho <= exp_biased) {
+      // ρ is less-than/equal to the exponent with bias.
+
+      // Diff between the biased exponent and ρ.
+      // Value in paper is spelled "e - e_min - ρ"
+      const uint32_t exp_diff = exp_biased - (uint32_t) rho;
+      // cMax * (exp_diff)
+      const mlib_int128 cmax_scaled =
+         mlib_int128_mul (cMax, MLIB_INT128_CAST (exp_diff));
+      // coeff * 10^rho * cMax * (exp_biased - rho)
+      result = mlib_int128_add (coeff_scaled, cmax_scaled);
+   } else {
+      const mlib_int128 biased_scale = mlib_int128_pow10 ((uint8_t) exp_biased);
+      result = mlib_int128_mul (biased_scale, coeff);
+   }
+
+   // Always add 2^127:
+   result = mlib_int128_add (result, i128_2pow127);
+
+   if (isNegative) {
+      // We calculated the value of the positive coefficient, but the decimal is
+      // negative. That's okay: Just flip the sign of the encoded result:
+      result = mlib_int128_negate (result);
```

Comment:
While refactoring, this was the natural result of folding operations in on each other. It made sense at the time, but I now had to re-derive why this worked:

The uint128 range is from `[0, 2^128)` with a midpoint at `2^127`. The "part1" of the mapping generates a value in the first half of that range. Adding `2^127` shifts the value into the upper half of the range. But this is only half of the range covered, since both `-N` and `+N` will map to the same point in the half of the range. Negating a two's-complement bit-pattern `S` is `(~S)+1`, but if you interpret `S` as unsigned, this just maps `S` to its reflection across the midpoint of the range. Thus, all negative numbers will be mapped to less-than the mapped positives, and greater-magnitude negatives map less-than lesser-magnitude negatives.

Reply:
Thank you for the explanation.